test/e2e/common.go

```diff
@@ -0,0 +1,370 @@
+	klusterletOperator           = "klusterlet"
+)
+
+var (
+	kubeClient         kubernetes.Interface
+	clusterCfg         *rest.Config
+	operatorClient     operatorclient.Interface
+	clusterClient      clusterclient.Interface
+	workClient         workv1client.Interface
+	bootstrapHubSecret *corev1.Secret
+)
+
+func createKlusterlet(name, clusterName, agentNamespace string) (realClusterName string) {
+	var klusterlet = &operatorapiv1.Klusterlet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Spec: operatorapiv1.KlusterletSpec{
+			RegistrationImagePullSpec: "quay.io/open-cluster-management/registration",
+			WorkImagePullSpec:         "quay.io/open-cluster-management/work",
+			ExternalServerURLs: []operatorapiv1.ServerURL{
+				{
+					URL: "https://localhost",
+				},
+			},
+			ClusterName: clusterName,
+			Namespace:   agentNamespace,
+		},
+	}
+
+	Expect(name).NotTo(BeEmpty())
+
+	// create agentNamespace and bootstrap-hub-kubeconfig secret
+	if agentNamespace != "" && agentNamespace != klusterletDefaultNamespace {
+		_, err := kubeClient.CoreV1().Namespaces().Get(context.TODO(), agentNamespace, metav1.GetOptions{})
+		if err != nil && errors.IsNotFound(err) {
+			namespace := v1.Namespace{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: agentNamespace,
+				},
+			}
+			_, err := kubeClient.CoreV1().Namespaces().Create(context.TODO(), &namespace, metav1.CreateOptions{})
+			Expect(err).NotTo(HaveOccurred())
+		}
+
+		secret := bootstrapHubSecret.DeepCopy()
+		secret.ObjectMeta.ResourceVersion = ""
+		secret.SetNamespace(agentNamespace)
+		_, err = kubeClient.CoreV1().Secrets(agentNamespace).Create(context.TODO(), secret, metav1.CreateOptions{})
+		Expect(err).NotTo(HaveOccurred())
+	}
+
+	// create klusterlet CR
+	realKlusterlet, err := operatorClient.OperatorV1().Klusterlets().Create(context.TODO(), klusterlet, metav1.CreateOptions{})
+	Expect(err).NotTo(HaveOccurred())
+	realClusterName = realKlusterlet.Spec.ClusterName
+
+	// the managed cluster should be created
+	Eventually(func() error {
+		clusters, err := clusterClient.ClusterV1().ManagedClusters().List(context.TODO(), metav1.ListOptions{})
+		if err != nil {
+			return err
+		}
+		if len(clusters.Items) == 0 {
+			return fmt.Errorf("there is no managed cluster created")
+		}
+		if len(clusters.Items) > 1 {
+			return fmt.Errorf("there are %v managed clusters created: %v", len(clusters.Items), clusters.Items)
+		}
+
+		if realClusterName != "" && realClusterName != clusters.Items[0].Name {
+			return fmt.Errorf("the managed cluster %v is not the one created %v", clusters.Items[0].Name, realClusterName)
+		}
+
+		if realClusterName == "" {
+			realClusterName = clusters.Items[0].Name
+		}
+
+		return nil
+	}, eventuallyTimeout*4, eventuallyInterval*4).Should(Succeed())
+
+	// approve csr
+	approveCSR(realClusterName)
+
+	// accept client
+	acceptsClient(realClusterName)
+	return
+}
+
+func approveCSR(clusterName string) {
```

Comment:
this should get injected w/ its dependencies (client etc)

Reply:
done with object `Tester`